### PR TITLE
Make it easier to reference models' views 

### LINKED ISSUE #12625: export_png doesn't find chrome webdriver because it was installed as chromedriver-binary
### Software versions

Python version      :  3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:36:39) [GCC 10.4.0]
IPython version     :  8.6.0
Tornado version     :  6.2
Bokeh version       :  3.0.2
BokehJS static path :  /home/jet08013/anaconda3/envs/python310/lib/python3.10/site-packages/bokeh/server/static
node.js version     :  v14.15.4
npm version         :  6.14.10
Operating system    :  Linux-5.15.0-52-generic-x86_64-with-glibc2.31

### Browser name and version

Google Chrome 107.0.5304.110

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

```python
import numpy as np
from bokeh.plotting import figure
from bokeh.io import export_png

f=figure()
x=np.linspace(-10,10,100)
y=x**2
f.line(x=x,y=y)
export_png(f,filename='out.png')
```
should generate a file 'out.png'.

### Observed behavior

The program crashed with the RunTime error:

```
Traceback (most recent call last):
  File "/home/jet08013/bokehexport.py", line 11, in <module>
    export_png(f,filename='out.png')
  File "/home/jet08013/anaconda3/envs/python310/lib/python3.10/site-packages/bokeh/io/export.py", line 116, in export_png
    image = get_screenshot_as_png(obj, width=width, height=height, driver=webdriver, timeout=timeout, state=state)
  File "/home/jet08013/anaconda3/envs/python310/lib/python3.10/site-packages/bokeh/io/export.py", line 254, in get_screenshot_as_png
    web_driver = driver if driver is not None else webdriver_control.get()
  File "/home/jet08013/anaconda3/envs/python310/lib/python3.10/site-packages/bokeh/io/webdriver.py", line 135, in get
    self.current = self.create()
  File "/home/jet08013/anaconda3/envs/python310/lib/python3.10/site-packages/bokeh/io/webdriver.py", line 139, in create
    driver = self._create(kind)
  File "/home/jet08013/anaconda3/envs/python310/lib/python3.10/site-packages/bokeh/io/webdriver.py", line 157, in _create
    raise RuntimeError("Neither firefox and geckodriver nor a variant of chromium browser and " \
RuntimeError: Neither firefox and geckodriver nor a variant of chromium browser and chromedriver are available on system PATH. You can install the former with 'conda install -c conda-forge firefox geckodriver'.
```

**My own investigations show that chromedriver was installed by conda under the name chromedriver-binary and
adding a symlink in the bin directory from chromedriver-binary to chromedriver fixed the problem.**

### Example code

```Python
import numpy as np
from bokeh.plotting import figure
from bokeh.io import export_png

f=figure()
x=np.linspace(-10,10,100)
y=x**2
f.line(x=x,y=y)
export_png(f,filename='out.png')
```

### Stack traceback or browser console output

Traceback (most recent call last):
  File "/home/jet08013/bokehexport.py", line 11, in <module>
    export_png(f,filename='out.png')
  File "/home/jet08013/anaconda3/envs/python310/lib/python3.10/site-packages/bokeh/io/export.py", line 116, in export_png
    image = get_screenshot_as_png(obj, width=width, height=height, driver=webdriver, timeout=timeout, state=state)
  File "/home/jet08013/anaconda3/envs/python310/lib/python3.10/site-packages/bokeh/io/export.py", line 254, in get_screenshot_as_png
    web_driver = driver if driver is not None else webdriver_control.get()
  File "/home/jet08013/anaconda3/envs/python310/lib/python3.10/site-packages/bokeh/io/webdriver.py", line 135, in get
    self.current = self.create()
  File "/home/jet08013/anaconda3/envs/python310/lib/python3.10/site-packages/bokeh/io/webdriver.py", line 139, in create
    driver = self._create(kind)
  File "/home/jet08013/anaconda3/envs/python310/lib/python3.10/site-packages/bokeh/io/webdriver.py", line 157, in _create
    raise RuntimeError("Neither firefox and geckodriver nor a variant of chromium browser and " \
RuntimeError: Neither firefox and geckodriver nor a variant of chromium browser and chromedriver are available on system PATH. You can install the former with 'conda install -c conda-forge firefox geckodriver'.

### Screenshots

_No response_

### LINKED ISSUE #13350: Possibility to reset plot from Python
### Problem description

Would like to be able to reset plot like done with the toolbar reset button from Python

Discuss on discord few years ago suggested Javascript alternative which I cannot make work (see below).
However still think having a pure Python method would be a nice alternative as it is not fully clear to me how I could reset one single specific plot on a multi-plot page using that alternative.

### Feature description

Function on a python plot which would trigger a plot view reset

### Potential alternatives

There is a discussion on discord on this: https://discourse.bokeh.org/t/how-to-reset-figure-directly-from-python-code-like-the-resettool-button/4787

Solution provided in discord does not work for me today.

With bokeh  2.4.3
document.querySelectorAll('.bk-tool-icon-reset[title="Reset"]') return correct list of reset tools items
However nothing happen on d.click()

With  bokeh 3.2.0
document.querySelectorAll('.bk-tool-icon-reset[title="Reset"]') return nothing
Have limited knowledge of javascript, but can see title parameter has been removed so I tried
document.querySelectorAll('.bk-tool-icon-reset'), still the same return nothing

### Additional information

Need for this reset option is linked to this issue when using bokeh plots in panel Tabs widget, I reported here:
https://github.com/holoviz/panel/issues/5298

I am looking for ways to mitigate the issue.

### PR DESCRIPTION
**Pull request by mattpap**
This makes `Bokeh.index` a `ViewManager` with a compatibility layer, so that `Bokeh.index[id]` still works. I also added access to the `index` to `CustomJS` callbacks via a new `context` (or `cb_context`) argument. Having such context is useful, because it allows users to avoid using globals, `Bokeh` specifically, which either may not be available (e.g. in ESM) or can be confusing when loading multiple instances of Bokeh on a page.

Due to issues with testing in my new setup, I had to resolve issue #12625.

TODO:

- [x] tests

addresses #13350 
fixes #12625

### CODE REVIEW

File: src/bokeh/settings.py [hunk 1]
```diff
@@ -567,6 +567,13 @@ def dev(self) -> bool:
     See the :class:`~bokeh.resources.Resources` class reference for full details.
     """)
 
+    chromedriver_path: PrioritizedSetting[str | None] = PrioritizedSetting("chromedriver_path", "BOKEH_CHROMEDRIVER_PATH", default=None, help="""
+    The name of or full path to chromedriver's executable.
+
+    This is used to allow ``bokeh.io.export`` to work on systems that use a different name
+    for ``chromedriver``, like ``chromedriver-binary`` or ``chromium.chromedriver`` (Snap).
```
Comment (bryevdv) [inline src/bokeh/settings.py]: If I didn't know what "Snap" was already, this parenthetical would be confusing at best. Suggest either expanding it with more context, or removing it.

File: src/bokeh/models/callbacks.py [hunk 2]
```diff
@@ -140,15 +141,23 @@ def __init__(self, *args, **kwargs) -> None:
...
     where ``args`` is a key-value mapping of user-provided parameters, ``obj``
-    refers to the object that triggered the callback, and ``data`` is a key-value
-    mapping of optional parameters provided by the caller.
+    refers to the object that triggered the callback, ``data`` is a key-value
+    mapping of optional parameters provided by the caller, and ``context`` is
+    additional context.
 
     This function can be an asynchronous function (``async function``).
+
+    The addtional context is composed of the following members:
```
Comment (bryevdv) [inline src/bokeh/models/callbacks.py]: suggest moving this block about context immediately after the "and context is additional context" sentence for continuity, and move the sentence about async after that, at the end.

File: src/bokeh/models/callbacks.py [hunk 3]
```diff
@@ -130,8 +130,9 @@ def __init__(self, *args, **kwargs) -> None:
     The code is made into the body of a function, and all of of the named objects in
     ``args`` are available as parameters that the code can use. Additionally,
     a ``cb_obj`` parameter contains the object that triggered the callback
-    and an optional ``cb_data`` parameter that contains any tool-specific data
-    (i.e. mouse coordinates and hovered glyph indices for the ``HoverTool``).
+    an optional ``cb_data`` parameter that contains any tool-specific data
+    (i.e. mouse coordinates and hovered glyph indices for the ``HoverTool``)
+    an additional context in ``cb_context`` argument.
```
Comment (bryevdv) [inline src/bokeh/models/callbacks.py]:
```suggestion
    and additional document context in ``cb_context`` argument.
```

I think it might make sense to call this specifically "document context" throughout? That seems to fit and would be less generic than "context" which could mean almost anything.

File: examples/interaction/js_callbacks/reset_plot.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    size=20,
+)
+
+reset = Button(label="Reset plot")
+reset.js_on_click(CustomJS(args=dict(plot=p), code="""
+export default ({plot}, _button, _data, {index}) => {
+    console.log(`Resetting ${plot}`)
+    const plot_view = index.get_one(plot)
+    plot_view.reset()
+}
```
Comment (bryevdv) [inline examples/interaction/js_callbacks/reset_plot.py:22]: what's the purpose of `export default` here? Seems like something we ought to discuss in documentation if it's a useful or advised pattern for users. We also need to discuss the new `context` in the narrative documentation as well. If you want to leave that for another PR, please open an issue regarding these needs.

File: bokehjs/src/lib/document/document.ts [hunk 5]
```diff
@@ -248,13 +251,15 @@ export class Document implements Equatable {
     return this._roots
   }
 
-  protected _add_root(model: HasProps): boolean {
-    if (includes(this._roots, model))
+  protected _add_root(...models: HasProps[]): boolean {
```
Comment (bryevdv) [inline bokehjs/src/lib/document/document.ts]: since this is private why not pluralize to `_add_roots` to be more descriptive of the new signature

**Review by bryevdv — COMMENTED** [review]
Really nice changes, just a few comments.

**Comment by bryevdv** [comment]
@mattpap just noting that #13350 was specifically about resetting *from Python*. I am not sure that's actually attained by this PR? (I think we'd need a new protocol event that could be triggered from Python to invoke the reset on the JS side)

Edit: I am not opposed to closing the issue with this PR, but I think we should explain the distinction and change in plans and why this is probably all that is needed, in the issue itself.

**Comment by bryevdv** [comment]
@mattpap as long it's now straightforward to expose `reset` in this way what you do think about exposing an explicit "request redraw" for users? Current users will try to do things like `source.change.emit()` to force a redraw (saw it again today)

File: examples/interaction/js_callbacks/reset_plot.py (hunk 4 above)
Comment (mattpap) [reply examples/interaction/js_callbacks/reset_plot.py:22]: `export default` is a part of ESM support that was added in PR #12812, which allows users to have explicit access to function's signature:
```ts
export default ({plot}, _button, _data, {index}) => {
  ...
}
```

Comment (bryevdv) [reply examples/interaction/js_callbacks/reset_plot.py:22]: OK, looks like we did not add in docs at that time, so I'd advise making an issue to expand docs in this area (unless you want to add them in this PR)

Comment (mattpap) [reply examples/interaction/js_callbacks/reset_plot.py:22]: I will do that separately. I started issue https://github.com/bokeh/bokeh/issues/13381 for that.

**Comment by mattpap** [comment]
> in this way what you do think about exposing an explicit "request redraw" for users?

`request_render()` is the API for this purpose:
```js
export default ({plot}, _target, _data, {index}) => {
    const plot_view = index.get_one(plot)
    plot_view.request_render()
}
```

**Comment by mattpap** [comment]
> just noting that https://github.com/bokeh/bokeh/issues/13350 was specifically about resetting from Python. I am not sure that's actually attained by this PR? (I think we'd need a new protocol event that could be triggered from Python to invoke the reset on the JS side)

Yeah, I so far overlooked the fact that we don't have any sensible programmatic way of running `CustomJS` (not attached to any tools or buttons or so).

**Comment by mattpap** [comment]
I will handle supporting running `CustomJS` independently of UI controls in another PR.

**Review by bryevdv — APPROVED** [review]
